### PR TITLE
ci: add Flutter test timing and reduce PR check overhead

### PR DESCRIPTION
**Pull request by ODevStudio**
## Summary

- Problem: PR checks are becoming slow as the Flutter suite has grown beyond 2,800 tests. The workflow also repeated setup/logging work and did not expose where test time was spent.
- Why it matters: the separate format job serialized larger jobs, repeated implicit dependency resolution added overhead, and expanded logs were difficult to use for later evidence-based optimization.
- What changed: format checking now covers only added/copied/modified/renamed Dart files inside `Analyze / test`; Flutter machine events are summarized and retained; Python tool tests are path-gated; two template tests were deleted.
- What did NOT change: the full Flutter suite and Linux release build smoke still run, with no selective test execution, sharding, smoke suite, or meaningful regression-test removal.

## Change Type (select all)

- [ ] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [x] Chore / infra
- [ ] Plugin (DYE2 or bundled skin)

## Scope (select all touched areas)

- [ ] BLE transport / device comms
- [ ] REST API / handlers
- [ ] WebSocket API
- [ ] Machine state / shot logic
- [ ] Scale / weight / flow
- [ ] Profiles / beans / grinders / workflows
- [ ] WebUI skins
- [ ] Plugins / JS runtime
- [ ] UI / Flutter widgets
- [ ] Storage / Drift database
- [x] CI / build / infra
- [ ] Docs / specs

## Linked Issues

- N/A

## Root Cause (if bug fix)

N/A - CI infrastructure change.

## Regression Test Plan (if bug fix or refactor)

N/A - the new parser has focused standard-library unit tests covering counts, failures, skips, hidden events, incomplete timing, unknown events, malformed lines, path normalization, Markdown escaping, and reporting failures.

## Documentation Obligations (required)

- [ ] API spec updated: `assets/api/rest_v1.yml` or `assets/api/websocket_v1.yml` (if REST/WebSocket changed)
- [ ] API docs updated: `doc/Api.md` (if user-facing endpoint changed)
- [ ] Plugin docs updated: `doc/Plugins.md` (if events/API changed)
- [ ] Skin docs updated: `doc/Skins.md` (if skin behavior changed)
- [ ] Profile docs updated: `doc/Profiles.md` (if profile handling changed)
- [ ] Device docs updated: `doc/DeviceManagement.md` (if device flows changed)
- [x] N/A - no docs affected

## Security Impact (required)

- New or changed REST endpoints? `No`
- New or changed WebSocket topics? `No`
- New or changed network calls? `Yes, CI only: one narrowly scoped GitHub artifact upload`
- BLE/USB surface changed? `No`
- File system access changed? `Yes, CI only: the parser reads the generated event file and appends to GITHUB_STEP_SUMMARY`
- Plugin sandbox boundary changed? `No`
- Risk and mitigation: the artifact path is exact, retention is seven days, no build directory is uploaded, and test names/paths are escaped before Markdown rendering.

## User-Visible Changes

None.

## Problem

PR checks are becoming slow as the Flutter suite has grown beyond 2,800 tests. The existing CI also has avoidable setup/logging overhead and provides poor visibility into which tests consume the time.

## What changed

- Removed the standalone `Format` job and its dependencies so `Analyze / test` and `Linux build smoke` can start independently.
- Added a null-delimited PR base/head Dart format check with full checkout history.
- Added `--no-pub` to analyze, test, and Linux build commands after explicit `flutter pub get`.
- Replaced expanded test output with `--machine`, `pipefail`, and `build/test-results/flutter-events.json` capture.
- Added a standard-library timing parser that writes counts, failures, total duration, and top-20 file/test timings to `GITHUB_STEP_SUMMARY` without masking Flutter's exit status.
- Added a seven-day `flutter-test-events` artifact upload that runs after failures and ignores a missing file.
- Added a native `git diff` path gate for `tools/**`, `tool/ci/**`, and this workflow; the Python job runs both tool test scripts only when relevant.
- Deleted only `test/unit_test.dart` and `test/widget_test.dart`, which were Flutter template examples.
- Left Flutter on `channel: stable` because the repository has no canonical version pin.
- Left DYE2 unchanged because `scripts/fetch_dye2_plugin.sh` already pins v0.1.4 and its SHA-256.

## What did not change

- Full Flutter tests still run.
- Linux release/build smoke behavior is unchanged.
- No selective test execution was introduced.
- No test sharding was introduced.
- No meaningful regression coverage was removed.

### Local gates (run before pushing)

- [x] Changed-Dart format command - no added/modified Dart files; the exact null-delimited command exited 0.
- [x] `flutter analyze --no-pub` - no issues found.
- [x] `flutter test` - official Ubuntu CI passed all 2,869 tests.
- [x] `./scripts/fetch_dye2_plugin.sh` - official Ubuntu CI passed; locally the download and SHA verification completed before Git Bash reported its missing `jq` executable.
- [x] `python3 tools/analyze_weight_flow_test.py` - 8 tests passed.
- [x] `python3 tool/ci/summarize_flutter_tests_test.py` - 3 tests passed.
- [x] Workflow YAML parsed and job dependencies were asserted.
- [x] A simulated failing producer remained non-zero through `tee` with `pipefail`.
- [x] A real machine event artifact and `GITHUB_STEP_SUMMARY` were produced.
- [x] The path gate returned `python=true` for this PR and `python=false` for a real unrelated Dart change.

### Manual verification (if applicable)

- OS / platform tested: Windows 11 locally and Ubuntu through GitHub Actions.
- Simulated devices? (`simulate=1`): `No`
- Real hardware? (DE1/Bengle/scale): `No`
- What was personally verified: parser output against local and official Ubuntu machine streams, artifact creation/retention, summary content, failure reporting, pipeline exit preservation, YAML structure, analyzer, Python tests, full Ubuntu Flutter tests, and Linux build smoke.
- Edge cases checked: skipped/hidden tests, incomplete timing, unknown event types, valid Flutter tool envelope events, malformed lines, missing event files, spaces/Markdown metacharacters, and absolute paths.
- What was not verified locally: a clean Windows full suite because QuickJS's Windows DLL was unavailable to the test process and one existing schema test is CRLF-sensitive. Ubuntu CI passed.

### Evidence

- [x] Test output: 8 existing Python tool tests and 3 parser tests passed; analyzer reported no issues.
- [x] Log snippets: the simulated producer returned non-zero through `tee`.
- [ ] Screenshot / recording (UI changes)
- [ ] curl / websocat output (API changes)
- Official Ubuntu summary: 2,869 successful, 0 failed, 0 skipped, 11.92 minute execution duration; the parser emitted top-20 file and individual test tables.
- Official raw artifact: 1,627,671 bytes uncompressed, 158,406 bytes compressed, retained through 2026-08-17 (seven days).
- Job ordering: `Analyze / test` and `Linux build smoke` both started at 10:50:37Z; build smoke finished at 10:53:52Z and analyze/test at 11:04:59Z.
- No percentage speed improvement is claimed; comparable GitHub Actions runs are not yet available.

## Compatibility & Migration

- Backward compatible? `Yes`
- Config / env changes needed? `No`
- Database migration needed? `No`

## Risks & Mitigations

- Risk: future Flutter machine events change shape.
  Mitigation: unknown events and unavailable optional timing are tolerated, parser failure cannot replace the Flutter exit status, and raw events remain downloadable.
- Risk: the Python job is skipped for a relevant change.
  Mitigation: the gate includes all of `tools/**`, `tool/ci/**`, and the PR workflow itself; a diff error fails the detector job rather than reporting no change.

### CODE REVIEW

File: tool/ci/summarize_flutter_tests.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            continue
+        try:
+            event = json.loads(line)
+        except (json.JSONDecodeError, TypeError):
+            malformed_lines += 1
+            continue
+        if not isinstance(event, dict):
+            continue
+
+        event_type = event.get("type")
```
Comment (tadelv) [inline tool/ci/summarize_flutter_tests.py:54]: Please preserve the actual failure diagnostics from Flutter's machine stream. The machine reporter emits assertion/error text and stack traces as separate `error` events keyed by `testID`; this parser currently ignores those events and only records the later `testDone(result: failure)`. Because the workflow also replaces `--reporter=expanded` with raw `--machine` output, a failed CI run becomes significantly harder to read: the summary identifies the failing test but not why it failed, while the job log is JSON. Please associate `error` events with their test and include at least the error message (and ideally a bounded/collapsible stack trace) in the summary. Add a parser test containing a real `error` event so this behavior is covered.

**Review by tadelv — CHANGES_REQUESTED** [review]
The CI/timing approach looks sound overall. I found one blocking issue: switching from the expanded reporter to raw `--machine` output currently drops the human-readable failure diagnostics unless the timing parser preserves them. Please address the inline comment; no broader architectural changes are requested for this PR.

**Review by tadelv — APPROVED** [review]
